php/libraries/NDB_Page.class.inc

```diff
@@ -927,6 +927,110 @@ class NDB_Page implements RequestHandlerInterface
+    ): array {
+        $projectsOptions = [];
+        $userProjectsIDs = \NDB_Factory::singleton()->user()->getProjectIDs();
+        foreach ($userProjectsIDs as $projectID) {
+            $projectsOptions[strval($projectID)]
+                = \Project::getProjectFromID($projectID)->getName();
+        }
+
+        if (!$numeric_keys) {
+            foreach ($projectsOptions as $id => $name) {
+                unset($projectsOptions[$id]);
+                $projectsOptions[$name] = $name;
+            }
+        }
+
+        return $projectsOptions;
+    }
+
+    /**
+     * Get the list of subprojects for which the current user has permission
+     * To populate the module subprojects filter options
+     *
+     * @param bool $numeric_keys If the returned subprojects keys
+     *                           should be numeric (SubprojectID)
+     *                           or strings (subproject name)
+     *
+     * @return array - an associative array of the form (ID => name)
+     */
+    public function getSubprojectOptions(
+        bool $numeric_keys = true
+    ): array {
+        $subprojectsOptions = [];
+        $userProjectsIDs    = \NDB_Factory::singleton()->user()->getProjectIDs();
+        foreach ($userProjectsIDs as $projectID) {
+            $subprojects = \Project::getProjectFromID($projectID)->getSubprojects();
+            foreach ($subprojects as $subproject) {
+                $subprojectId = $subproject['subprojectId'];
+                if (!isset($subprojectsOptions[$subprojectId])) {
+                    $subprojectsOptions[$subprojectId] = $subproject['title'];
+                }
+            }
+        }
+
+        if (!$numeric_keys) {
+            foreach ($subprojectsOptions as $id => $name) {
+                unset($subprojectsOptions[$id]);
+                $subprojectsOptions[$name] = $name;
+            }
+        }
+
+        return $subprojectsOptions;
+    }
+
+    /**
+     * Get the list of sites for which the current user has permission
+     * To populate the module sites filter options
+     *
+     * @param bool  $numeric_keys       If the returned sites keys
+     *                                  should be numeric (CenterID)
+     *                                  or strings (site name)
+     * @param array $allsitesPermission A list of permission
+     *                                  which gives access to all sites
+     * @param bool  $study_site         If the returned sites
+     *                                  should be study sites only
+     *
+     * @return array An associative array of the form (ID => name)
+     */
+    public function getSiteOptions(
+        bool $numeric_keys = true,
+        array $allsitesPermission = ['access_all_profiles'],
+        bool $study_site = true
+    ): array {
+        $user         = \NDB_Factory::singleton()->user();
+        $site_options = $user->getStudySites();
+
+        foreach ($allsitesPermission as $permission) {
```

Comment:
This one's a little more complicated since it incorporates the site parameter, but given that our menu filters should all be extending `DataFrameworkMenu` where there's a `abstract public function allSitePermissionNames() : ?array;` function I don't think that's a problem. This could still be moved there instead of the NDB_Page class.